noria-server/mir/src/optimize.rs

```diff
@@ -1,19 +1,217 @@
+                                                            if *j == i {
+                                                                candidate = false;
+                                                            }
+                                                        }
+                                                    }
+                                                },
+                                                _ => {},
+                                            }
+                                        },
+                                        None => {},
+                                    }
+                                }
+                            }
+                        },
+                        _ => {},
+                    };
+                }
+            },
+            _ => {},
+        };
+        if candidate {
+            candidate_nodes.push(n);
+        }
+    }
+
+    // 3. For each candidate, merge it, and update all parents/children
+    // of the newly merged node.
+
+    let mut new_nodes = Vec::new();
+
+    for n in candidate_nodes {
+        let temp = n.borrow();
+        let child = temp.children.first().unwrap().borrow();
+        // determine which is which
+        // (this was relevant when we supported either order of agg/filter,
+        // but I'm leaving it in place in case we resume that)
+        let (agg, filter) = match n.borrow().inner {
+            MirNodeType::Aggregation { .. }  => (n.borrow().inner.clone(), child.inner.clone()),
+            MirNodeType::Filter { .. } => (child.inner.clone(), n.borrow().inner.clone()),
+            _ => unreachable!(),
+        };
+
+        let mut new_name = child.name.clone();
+        new_name.push_str("_filteragg");
+
+        let new_node = MirNode::new(
+            &new_name,
+            child.from_version,
+            child.columns.clone(),
+            MirNodeType::FilterAggregation {
+                on: match &agg {
+                    MirNodeType::Aggregation { on, group_by: _, kind: _ } => on.clone(),
+                    _ => unreachable!(),
+                },
+                else_on: None,
+                group_by: match &agg {
+                    MirNodeType::Aggregation { on: _, group_by, kind: _ } => group_by.to_vec(),
+                    _ => unreachable!(),
+                },
+                //group_by.into_iter().cloned().collect(),
+                kind: match &agg {
+                    MirNodeType::Aggregation { on: _, group_by: _, kind } => {
+                        match kind {
+                            Aggregation::COUNT => FilterAggregation::COUNT,
+                            Aggregation::SUM => FilterAggregation::SUM,
+                        }
+                    },
+                    _ => unreachable!(),
+                },
+                conditions: match &filter {
+                    MirNodeType::Filter { ref conditions } => conditions.to_vec(),
```

Comment:
Hmm, I wonder if `conditions` needs some adjustment here depending on whether the original filter was before or after given that the aggregation moves columns around?

Reply:
Great point. Will think about this more soon.

Reply:
It looks like the reason this was working before was because when merging {parent, child} I was giving the new node child.columns for its column vec. I think instead it should be getting parent.columns and adjusting the filter indices as you suggested.

...Actually, now I'm blanking. Dumb question: are the columns the input or output columns of the node? (Either way I think it needs the fix you suggested.)

Reply:
I believe columns is the output, but I don't know what MIR does here, so we'd have to check with @ms705.

Reply:
Yup, just confirmed that with Malte. Will finish fixing this now.

Reply:
OK, I have written some code that I believe will catch this case.

I believe previously all the tests were passing because they were Aggregation followed by Filter, and the column reordering would only cause a problem with Filter followed by Aggregation. I am having trouble coming up with a simple test case that forces it to be in that order. The obvious way to write a query that does this would be to aggregate on the column being filtered, because then you have to do the filter first, but then I realized this case can't merge the two operators anyway because the filter column doesn't exist in the output columns and thus there's no way to represent it. (This is an interesting limitation of our current filter model.)

Anyway, now there's *also* code to make sure we're not merging in that case, and a test for that, but no test for the original problem you spotted. Feedback on the new code would be much appreciated.

Reply:
Never mind, needs another rewrite -- Malte just pointed out that, although the columns field is the output, the filter indices are over the input (i.e. the parent's columns field). 

Oh, well -- at least I understand the code base better now.

Reply:
It is probably also okay for the code to just _detect_ this case and error, rather than coming up with the correct implementation for it, since it seems to be tricky to trigger.